p2p_beef_tx.go

```diff
@@ -0,0 +1,229 @@
+package paymail
+
+import (
+	"encoding/hex"
+	"errors"
+
+	"github.com/libsv/go-bt/v2"
+)
+
+type CompoundMerklePath []map[string]uint64
+
+type CMPSlice []CompoundMerklePath
+
+const (
+	BEEFMarkerPart1 = 0xBE
+	BEEFMarkerPart2 = 0xEF
+)
+
+const (
+	HasNoCMP = 0x00
+	HasCMP   = 0x01
+)
+
+type TxData struct {
+	Transaction *bt.Tx
+	PathIndex   *bt.VarInt
+}
+
+type DecodedBEEF struct {
+	CMPSlice CMPSlice
+	TxData   []TxData
+}
+
+func DecodeBEEF(beefHex string) (*DecodedBEEF, error) {
+	beefBytes, err := extractBytesWithoutVersionAndMarker(beefHex)
+	if err != nil {
+		return nil, err
+	}
+
+	cmpSlice, remainingBytes, err := DecodeCMPSliceFromStream(beefBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	transactions, err := DecodeTransactionsWithPathIndexes(remainingBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DecodedBEEF{
+		CMPSlice: cmpSlice,
+		TxData:   transactions,
+	}, nil
+}
+
+func DecodeCMPSliceFromStream(hexBytes []byte) (CMPSlice, []byte, error) {
+	if len(hexBytes) == 0 {
+		return nil, nil, errors.New("provided hexStream is empty")
+	}
+
+	nCMPs, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
+	hexBytes = hexBytes[bytesUsed:]
+
+	var cmpPaths []CompoundMerklePath
+	for i := 0; i < int(nCMPs); i++ {
+		cmp, bytesUsedToDecodeCMP, err := NewCMPFromStream(hexBytes)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cmpPaths = append(cmpPaths, cmp)
+		hexBytes = extractRemainingBytes(hexBytes, bytesUsedToDecodeCMP)
+	}
+
+	cmpSlice := CMPSlice(cmpPaths)
+
+	return cmpSlice, hexBytes, nil
+}
+
+func NewCMPFromStream(hexBytes []byte) (CompoundMerklePath, int, error) {
+	hexToProcess := make([]byte, len(hexBytes))
+	copy(hexToProcess, hexBytes)
```

Comment:
Why do you copy the slice?

Reply:
it was leftover from debugging code - thought it was a problem when I couldn't find an error - removed, thanks